[IDLE-261] feat: 소셜 로그인 이메일 조회 API 개발

0112leesy opened the pull request:
## 🧑‍💻 작업 사항

### 소셜 로그인 이메일 조회 API
환경설정 탭에서 소셜 로그인 정보를 조회할 수 있도록 이메일 조회 API를 구현했습니다.
(애플 소셜로그인 이슈 처리때문에 개발이 늦어져서 대단히 죄송합니다 😅)
- 요청
```
GET http://mybrary.kr/user-service/api/v1/users/{userId}/profile/email
```

- 응답
```
{
  "status": "200 OK",
  "message": "사용자의 로그인된 이메일 계정을 조회했습니다.",
  "data": {
    "email": "email@mybrary.com"
  }
}
```

### 소셜 로그인 실패 시 예외 처리 변경
- 기존
<img width="1290" alt="image" src="https://github.com/SWM-IDLE/mybrary-user-service/assets/78717113/bceb4c40-06c0-4f34-92b5-c42a354e8c3b">

- 응답 body로 errorCode, errorMessage 전달
- 모바일 환경에서, 웹뷰에 던져진 errorCode, errorMessage를 읽어올 수 없는 문제 발생
- 따라서 소셜 로그인 성공 시 accessToken, refreshToken을 전달하는 방법처럼,
- 예외 발생 시 errorCode, errorMessage를 redirect url의 파라미터로 전달되도록 수정

<br>


- 변경
<img width="1512" alt="image" src="https://github.com/SWM-IDLE/mybrary-user-service/assets/78717113/9104ce96-a693-40c2-a4ed-2d1df996be17">

- kr.mybrary://?errorCode={errorCode}&errorMessage={errorMessage} 형식으로 리턴
- 클라이언트에서 에러 파라미터를 잘 전달 받아서 예외 처리가 잘 수행되면 좋겠습니다!! 😄


<!-- 어떤 변경사항이 있는지, PR 타이틀과 동일하다면 Skip -->

<br><br>

## 🔗 링크

<!-- 관련된 대화가 이루어진 슬랙 링크 혹은 피그마 링크를 첨부. -->
<!-- - [프레이머](https://framer.com/projects/2--gSDcZoNKqU6dqCUQUYQe-53rEr?node=tQMWSSKqy)  -->

<br><br>

## 🐰 시급한 정도
🏃‍♂️ 보통 : 최대한 빠르게 리뷰 부탁드립니다. 
<!-- 🚨 긴급 : 선 어프루브 후 리뷰를 부탁드립니다. (리뷰는 추후 반영)  -->
<!-- 🐢 천천히 : 급하지 않습니다. -->

<br><br>

## 📖 참고 사항

<!-- 공유할 내용, 레퍼런스, 추가로 발생할 것으로 예상되는 이슈, 스크린샷 등을 넣어 주세요.  -->
<!-- 변경사항이 큰 경우 집중해야 할 부분, 또는 불안해서 봐주었으면 하는 부분 등 -->

0112leesy committed:
[IDLE-261] feat: 로그인된 소셜 이메일 조회 API 구현

  - 이메일이 조회되지 않을 경우 빈 문자열 반환

0112leesy committed:
[IDLE-261] test: 로그인된 소셜 이메일 조회 API 테스트 및 문서화

0112leesy committed:
[IDLE-261] feat: OAuth2LoginFailureHandler 수정

  - 기존: 응답 body에서 errorCode, errorMessage 전달
  - 문제: 웹뷰에서 리턴되는 응답 body를 모바일에서 읽어올 수 없음
  - 수정: redirect url로 errorCode, errorMessage 응답

0112leesy committed:
[IDLE-261] feat: 애플 소셜로그인 관련 예외 메시지 영어로 변경

  - redirect url로 errorMessage 파라미터를 넘길 때 한글이 인코딩 되는 문제 발생하여 에러 메시지를 영어로 변경
  - 관련 테스트 코드 수정

0112leesy committed:
[IDLE-261] feat: 애플 소셜로그인 예외 발생 시 redirect url로 errorCode, errorMessage 응답되도록 수정

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/user/domain/UserServiceImpl.java:
```diff
@@ -242,6 +242,23 @@ public ProfileImageUrlServiceResponse deleteProfileImage(ProfileImageUpdateServi
+    @NotNull
+    private String getUserEmail(String loginId) {
+        User user = getUser(loginId);
+        if(user.getEmail() == null) {
```
r3: user.getEmail() == null 비교 보다는, user 클래스 내에서 이메일이 존재하는지 판단하는 메서드가 있으면 좋을 것 같습니다.
```java
// User 클래스 내
public boolean hasEmail() {
    return this.email != null;
}

// getUserEmail 메서드
if (user.hasEmail()) {
    return "";
}

```

0112leesy replied:
넵넵! User 엔티티에 메서드를 추가해보겠습니다!

minnseong reviewed (approved):
너무 수고많으셨습니다 !!

DonggyuJin reviewed (approved):
소셜 로그인 예외 처리까지 신경써주셔서 감사합니다!! 고생하셨습니다 얼른 적용해보겠습니다 👍

0112leesy committed:
[IDLE-261] refactor: 코드 리뷰 반영

  - r3: user 클래스 내에서 이메일이 존재하는지 판단하는 메서드 추가
  - sonarcloud: isEqualTo("") 대신 isEmpty() 사용
